File: src/chain-observers/providers/vote-provider.ts
import { TAccountName, vote } from "@hiveio/wax";
import { WorkerBeeArrayIterable, WorkerBeeIterable } from "../../types/iterator";
import { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { IOperationTransactionPair, OperationClassifier } from "../classifiers/operation-classifier";
import { DataEvaluationContext } from "../factories/data-evaluation-context";
import { ProviderBase } from "./provider-base";

export type TVoteProvided<TAccounts extends Array<TAccountName>> = {
  [K in TAccounts[number]]: WorkerBeeIterable<IOperationTransactionPair<vote>>;
};

export interface IVoteProviderData<TAccounts extends Array<TAccountName>> {
  votes: Partial<TVoteProvided<TAccounts>>;
};

export interface IVoteProviderOptions {
  voters: string[];
}

export class VoteProvider<TAccounts extends Array<TAccountName> = Array<TAccountName>> extends ProviderBase<IVoteProviderOptions> {
  public readonly voters = new Set<TAccountName>();

  public pushOptions(options: IVoteProviderOptions): void {
    for(const account of options.voters)
      this.voters.add(account);
  }

  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [OperationClassifier]
  }

  public async provide(data: DataEvaluationContext): Promise<IVoteProviderData<TAccounts>> {
    const result = {
      votes: {}
    } as IVoteProviderData<TAccounts>;

    const operations = await data.get(OperationClassifier);
    if (operations.operationsPerType.vote_operation)
      for(const operation of operations.operationsPerType.vote_operation) {
        if(this.voters.has(operation.operation.voter) === false)
          continue;

        const account = operation.operation.voter;

        if (!result.votes[account])
          result.votes[account] = new WorkerBeeArrayIterable();

        const storage = result.votes[account] as WorkerBeeArrayIterable<IOperationTransactionPair<vote>>;
        storage.push({
          operation: operation.operation,
          transaction: operation.transaction
        });
      }

    return result;
  }
}

File: python/workerbee/chain_observers/providers/vote_provider.py
"""VoteProvider — provides vote operations grouped by voter.

Mirrors src/chain-observers/providers/vote-provider.ts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .._ordered_set import OrderedSet
from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
from ..classifiers.operation_classifier import OperationClassifier
from .provider_base import ProviderBase

if TYPE_CHECKING:
    from ..factories.data_evaluation_context import DataEvaluationContext
    from ..payloads import OpBodyTransactionPair, VotesPayload

VoteOptions = dict[str, list[str]]


class VoteProvider(ProviderBase):
    def __init__(self) -> None:
        self.voters: OrderedSet[str] = OrderedSet()

    def push_options(self, options: VoteOptions) -> None:
        voters = options.get("voters", [])
        for account in voters:
            self.voters.add(account)

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        return [OperationClassifier]

    async def provide(self, data: DataEvaluationContext) -> VotesPayload:
        result: dict[str, list[OpBodyTransactionPair]] = {}

        operations = await data.get(OperationClassifier)
        vote_ops = operations["operations_per_type"].get("vote_operation")
        if vote_ops:
            for operation in vote_ops:
                voter = operation["operation"]["voter"]
                if not isinstance(voter, str):
                    continue
                if voter not in self.voters:
                    continue

                if voter not in result:
                    result[voter] = []

                result[voter].append(
                    {
                        "operation": operation["operation"],
                        "transaction": operation["transaction"],
                    }
                )

        return {"votes": result}
